### PR TITLE
Fix undesired specialization on slice after split.

### PR DESCRIPTION
**Pull request by ysiraichi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142372
* #145681

Fix: #141251

This PR adds a few static guard checks when decomposing and lowering the `slice`
operation, so that we avoid adding unnecessary guards. Specifically, when clamping the end
values.

In summary, the changes are:

- `slice` dynamo decomposition: checks `end >= sizes[dim]` statically. If we don't know
  that, the following guard ensures that we (don't) need clamping.
- `evaluate_min` inductor `sizevar` function: checks whether we can solve it statically or
  not, before actually creating a new guard.

The latter had to be changed because `evaluate_min` (called by `ir.SliceView` constructor)
would always try to create a guard based on the hints operation result. However, if both
`left` and `right` hints were true, it would default to `left <= right` guard. By checking
the guards statically before, we can avoid that.

```python
N = 16

@torch.compile(backend="inductor", dynamic=False, fullgraph=True)
def fn(x):
    splits = torch.ops.aten.split.Tensor(x, N)
    first = splits[0]
    return torch.ops.aten.slice.Tensor(first, 0, 0, N)

x = torch.arange(N)
torch._dynamo.mark_dynamic(x, 0)

fn(x)
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: test/export/test_export.py [hunk 1]
```diff
@@ -9500,7 +9500,7 @@ def forward(self, x):
                 return x[:out_size]
 
         dim_min = 5
-        dim_max = 10
+        dim_max = 8
```
Comment (ysiraichi) [inline test/export/test_export.py:9503]: I had to do this because changing the slice comparison resulted in the following single-symbol guard: `round(((float(L['x'].size()[0])) / 2.0)) < L['x'].size()[0]`, which is obviously true. However, we can't get to that conclusion with only value range analysis, because we end up with: `[2, 5] < [5, 10]`.

**Comment by ezyang** [comment]
Is failing tests. Do you still want review?

File: test/export/test_export.py [hunk 2]
```diff
@@ -6177,7 +6177,7 @@ class M(torch.nn.Module):
             def forward(self, x, y):
                 a = x.item()
                 torch._check_is_size(a)
-                torch._check(a <= y.size(0))
+                torch._check(a < y.size(0))
```
Comment (ysiraichi) [inline test/export/test_export.py:6180]: I had to do this because with the former check, we couldn't tell whether `a >= y.size(0)`, creating a new guard.

Comment (ysiraichi) [reply test/export/test_export.py]: Reverted. Not needed after replacing the guards tweak with the static checks.

File: test/export/test_export.py (hunk 1 above)
Comment (ysiraichi) [reply test/export/test_export.py]: Reverted. Not needed after replacing the guards tweak with the static checks.

**Comment by ysiraichi** [comment]
@ezyang I think I found a better solution compared with what we had before. Basically, I added a few static guard checks before falling back to creating new guards. See the PR description for more details.

**Comment by ysiraichi** [comment]
@ezyang This PR is ready for review. Could you take a look at it whenever you have some time?

**Review by ezyang — APPROVED** [review]

**Comment by ysiraichi** [comment]
@laithsakka @ezyang Could you provide more information on what went wrong with this PR?

**Comment by ezyang** [comment]
I don't have a minimal repro, but here's the description:

> From the tlparse it says "Could not guard on data-dependent expression Eq(u45 + u46 - u51 - u52 - u53 - u54 - u55, 0), but if we looked into the log, we can find there is a guard: u51 + u52 + u53 + u54 + u55 + u56 == u45 + u46

So this should be relandable if we beef up reasoning to be able to do the inference here.

**Comment by ysiraichi** [comment]
Not sure I get it. But, here's what is on my mind:

- The already existing guard is actually a runtime assert added by the call of `expect_true`
- Then, at some point, we call `guard_bool` on a similar expression, which raises the data-dependent error

If this is the case, given `Eq(e0, e1)`, one solution I can think of is:

- Add the canonicalization (call `canonicalize_bool_expr`) of the following implications to `ShapeEnv.get_implications`: `Eq(e0, e1)` and `Eq(-e0, -e1)`
- Inside `ShapeEnv._maybe_evaluate_static`, canonicalize `expr` add a second axiom substitution trial `expr = canonicalize_bool_expr(expr).xreplace(subst)`

This should replace the whole expression with `sympy.true`.
@ezyang What do you think?

**Comment by laithsakka** [comment]
@ysiraichi
I am looking at this and will make a comment explaining what happened to discuss how we shall procceed

**Comment by laithsakka** [comment]
here is a repo of the issue that blocked landing this diff:

```
@torch.compile(backend="eager", fullgraph=True)
def cf_check(a, b, c):
    u0 = a.size()[0]
    u1 = b.size()[0]
    u2 = c.size()[0]
    torch._check(u0+u2==u1)

    # This fails.
    guard_size_oblivious(u1-u0==0)
    # But we know that u0 = u1-u2
    # Hence the guard above can become
    # guard_size_oblivious(u2==0)

    return torch.tensor(True)
a = torch.ones(10)
torch._dynamo.decorators.mark_unbacked(a, 0)
b = torch.ones(20)
torch._dynamo.decorators.mark_unbacked(b, 0)
c = torch.ones(10)
torch._dynamo.decorators.mark_unbacked(c, 0)
cf_check(a, b, c)
```

basically  removing the guard added before this diff causes 
guard_size_oblivious(u1-u0==0)
to fail. 
although if we do better reasoning, we probably can guard on it. 
note that u0 and u1 are size like. 
one thing that can be done in theory us replacing u0 with u1-u2 and then  guard_size_oblivious(u2==0) is False.

**Comment by laithsakka** [comment]
question in the original error why do not use instead maybe_mark_dyanamic and just not bather with it. 
or are you really trying to avoid some reocmpilations?

**Comment by ysiraichi** [comment]
Hi, Laith. Thank you for the repro.

> why do not use instead maybe_mark_dyanamic

Right. I think that can be a solution to the original error. But, I still think the code in question is valid, and we should support it (as is). 

So, I was thinking about implementing [my proposal](https://github.com/pytorch/pytorch/pull/142372#issuecomment-2554203774), which should solve this error. What do you think?

**Comment by laithsakka** [comment]
> Hi, Laith. Thank you for the repro.
> 
> > why do not use instead maybe_mark_dyanamic
> 
> Right. I think that can be a solution to the original error. But, I still think the code in question is valid, and we should support it (as is).
> 
> So, I was thinking about implementing [my proposal](https://github.com/pytorch/pytorch/pull/142372#issuecomment-2554203774), which should solve this error. What do you think?

to make sure I understand can you explain how that would work on the repo above to make that would make us infer
guard_size_oblivious(u1-u0==0) is False?

**Comment by laithsakka** [comment]
one issue with the proposed reasoning in the repo i put above is that when we call 
guard_size_oblivious(u1-u0==0) we are saying you can assume u1>=2 and u0>=2
but we are not saying that u2 also is >=2

so even if we reach:
guard_size_oblivious(u2==0)
that does not mean u2 >=2 hence we can not return false in this case.

from my understanding,  guard_size_oblivious apply the assumption on unbacked symints  in the input expressions only right? or is it safe to extend it to all?

**Comment by ysiraichi** [comment]
Hi, @laithsakka. Sorry for taking long to reply.
I thought about my proposal, and realize it wouldn't work. So, I thought of some simplification process that would work: use the runtime asserts as a knowledge-base for substituting the free variables.
I have pushed my changes into a PR #145681. Let me know what you think.
